File: gocp/internal/query/unanalyzable_gate_db_test.go
package query_test

import (
	"testing"

	"github.com/ridi-oss/proxy-monster/gocp/internal/dbtest"
	pb "github.com/ridi-oss/proxy-monster/gocp/internal/pb"
)

// UnanalyzableGateDbTest.kt — 67 LOC, 1 case (06-query-decision.md §7, step 16).
//
// The unanalyzable-statement gate (docs/facts-emission.md) end to end on real PostgreSQL + real Cedar.
// The analyzer cannot prove lineage for a NATURAL JOIN — shared-column lineage is ambiguous — so
// `resolved=false`, and decideQuery asks the DATASOURCE for the `sql.unanalyzable` exception:
//
//   - the production floor (no exception policy) → DENY, fail-closed; and
//   - a datasource that shipped a `sql.unanalyzable` permit → ALLOW, relaying the ORIGINAL statement
//     verbatim (passthrough, no masks) — the permissive development-datasource posture.
//
// The gate fires BEFORE the column/table/function gates (an unresolved probe emits no facts for them),
// so a missing read grant on the referenced tables is irrelevant here; this isolates the analyzable
// decision. That is also why the statement scans the UNGRANTED `orders` table without that mattering.
func TestUnanalyzableGateDb(t *testing.T) {
	fx := newEnforcementFixture(t, dbtest.EnginePostgres)

	// Admitted (a single SELECT, so it passes the sql.select kind gate) but unanalyzable: the probe
	// rejects NATURAL JOIN early, before catalog resolution, so this is `resolved=false` regardless of
	// the schema.
	const unanalyzable = "select count(*) from orders natural join users"

	// ⚠️ ONE FUSED CASE, and the fusing is deliberate — the Kotlin's comment says so in as many words.
	// The permit created in part 2 PERSISTS on the shared fixture, so the floor DENY must be observed
	// BEFORE it exists. Splitting these into two Go subtests would leave the order to whoever edits the
	// file next, and the failure mode is silent: part 1 would still "pass" as a DENY only until someone
	// reordered them, after which it would fail for a reason that looks like a decideQuery regression.
	// KT: UnanalyzableGateDbTest.kt#unanalyzable denies on the floor, then a sql-unanalyzable permit relays it verbatim
	t.Run("unanalyzable denies on the floor, then a sql-unanalyzable permit relays it verbatim", func(t *testing.T) {
		// 1. Production floor — no exception policy → deny-by-default (fail-closed).
		floor := gateDecide(fx, dbtest.FixturePrincipal, unanalyzable, nil)
		wantAction(t, floor, pb.EnfAction_DENY, "no sql.unanalyzable policy → deny-by-default")
		wantDetailContains(t, floor, "could not analyze", "the fail-closed reason is preserved")

		// 2. A datasource that shipped the exception (the permissive development posture) → relay
		//    verbatim. AddCedarPolicy also bumps the policy-store state version, without which the
		//    already-built CedarEngine would keep serving its cached pre-permit policy set.
		fx.AddCedarPolicy("test-dev-unanalyzable",
			`permit(principal, action == Action::"sql.unanalyzable", resource == Datasource::"`+fx.DatasourceName+`");`)

		permitted := gateDecide(fx, dbtest.FixturePrincipal, unanalyzable, nil)
		wantAction(t, permitted, pb.EnfAction_ALLOW, "sql.unanalyzable permit → relay the original statement verbatim")
		if !permitted.Passthrough {
			t.Error("an unanalyzable relay is a verbatim passthrough (no rewrite, no masks)")
		}
		if len(permitted.Masks) != 0 {
			t.Errorf("no masks are applied to an unanalyzable relay, got %d", len(permitted.Masks))
		}
		wantDetailContains(t, permitted, "sql.unanalyzable", "the ALLOW is attributed to the exception")
	})
}

File: control-plane/src/test/kotlin/com/ridi/oss/proxymonster/controlplane/UnanalyzableGateDbTest.kt
package com.ridi.oss.proxymonster.controlplane

import com.ridi.oss.proxymonster.grpc.EnfAction
import com.ridi.oss.proxymonster.controlplane.authz.CedarPolicyInput
import com.ridi.oss.proxymonster.controlplane.support.EnforcementFixture
import com.ridi.oss.proxymonster.controlplane.support.requireDockerOrSkip
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * Unanalyzable-statement gate (docs/facts-emission.md) end-to-end on real PostgreSQL + real Cedar. The
 * analyzer cannot resolve an unknown output column → `resolved=false`.
 * decideQuery asks the datasource for the `exception.unanalyzable` exception:
 *  - production floor (no exception policy) → DENY; and
 *  - a datasource that shipped a `exception.unanalyzable` permit → ALLOW, relaying the ORIGINAL statement verbatim
 *    (passthrough, no masks) — the permissive development-datasource posture.
 * The gate fires before the column/table/function gates (an unresolved probe emits no facts for them), so a
 * missing read grant on the referenced tables is irrelevant here — this isolates the analyzable decision.
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class UnanalyzableGateDbTest {
    private lateinit var fx: EnforcementFixture
    private val principal = "analyst@example.com"

    // Admitted as SELECT but unresolved against the catalog.
    private val unanalyzable = "select missing_column from orders"

    @BeforeAll
    fun setup() {
        requireDockerOrSkip()
        fx = EnforcementFixture.postgres()
    }

    private fun decide(sql: String) = decideQuery(
        principal = principal, ds = fx.datasource, sql = sql, channel = Channel.WIRE,
        catalog = fx.datasourceStore.catalog(fx.datasource.id), policyStore = fx.policyStore, accessStore = fx.accessStore,
        userGroupStore = fx.userGroupStore, roleResolver = fx.roleResolver, authz = fx.authz,
    )

    // One ordered test: the floor DENY must be observed BEFORE the permit is created (the permit persists on
    // the PER_CLASS fixture), so the two assertions live in one method to fix their order.
    @Test
    fun `unanalyzable denies on the floor, then a sql-unanalyzable permit relays it verbatim`() {
        // 1. Production floor — no exception policy → deny-by-default (fail-closed).
        val floor = decide(unanalyzable)
        assertEquals(EnfAction.DENY, floor.action, "no exception.unanalyzable policy → deny-by-default")
        assertTrue(floor.detail?.contains("could not analyze") == true, "the fail-closed reason is preserved: ${floor.detail}")

        // 2. A datasource that shipped the exception (the permissive development posture) → relay verbatim.
        fx.cedarPolicyStore.create(
            CedarPolicyInput(
                name = "test-dev-unanalyzable",
                cedarSrc = """permit(principal, action == Action::"exception.unanalyzable", resource == Datasource::"${fx.datasource.name}");""",
            ),
            updatedBy = "test",
        )
        val permitted = decide(unanalyzable)
        assertEquals(EnfAction.ALLOW, permitted.action, "exception.unanalyzable permit → relay the original statement verbatim")
        assertTrue(permitted.passthrough, "an unanalyzable relay is a verbatim passthrough (no rewrite, no masks)")
        assertTrue(permitted.masks.isEmpty(), "no masks are applied to an unanalyzable relay")
        assertTrue(permitted.detail?.contains("exception.unanalyzable") == true, "the ALLOW is attributed to the exception: ${permitted.detail}")
        // Both verdicts must ask for a refetch first. A cross-schema statement whose qualifier this connection
        // never fetched is unresolvable for that reason alone, and reaches this gate indistinguishable from a
        // genuinely unanalyzable one — so without catalogMiss the relay stands on a catalog that never held
        // the schema, and every column the analyzer would have masked is relayed in the clear.
        assertTrue(floor.catalogMiss, "the deny carries catalogMiss so decideConnection refetches + retries first")
        assertTrue(permitted.catalogMiss, "the relay carries catalogMiss so refetch-first runs before it stands")
    }

    // The cross-schema case the flag exists for, driven through the REAL analyzer rather than a synthetic
    // failure: a statement qualifying a schema absent from the decision's catalog resolves nowhere, so it
    // lands on the unanalyzable gate carrying that schema as a refetch candidate.
    @Test
    fun `a statement naming an unheld schema asks for that schema before relaying`() {
        val candidate = "goods_store"
        // Assert the premise rather than assume it: were this schema ever added to the fixture, the statement
        // would resolve and the test would pass without exercising the gate at all.
        assertTrue(
            fx.datasourceStore.catalog(fx.datasource.id).none { it.schema == candidate },
            "the probe schema must be absent from the catalog for this test to mean anything",
        )
        val ctx = decide("select id from $candidate.orders")
        assertTrue(ctx.catalogMiss, "an unheld-schema statement must request a refetch, not stand on the partial catalog")
        assertTrue(
            candidate in ctx.schemaCandidates,
            "the unheld schema must be named for the refetch, got ${ctx.schemaCandidates}",
        )
    }
}
